Connection: Select One of Multiple Blog Tokens

mdawaffe opened the pull request:
In certain situations, it is useful to:
* Not rely on the database for the blog token, and
* Allow multiple blog tokens per site (for secret rotation, for example).

### Changes proposed in this Pull Request:
* Allow the blog token to be set via a new `JETPACK_BLOG_TOKEN` constant.
* Support multiple blog tokens in `Jetpack_Data::get_access_token()` by returning either the first or the one matching the provided `$token_key`.
* Never call `Jetpack_Options::get_option( 'blog_token' )` directly. Switch to `Jetpack_Data::get_access_token()`
* Since the blog token is used to sign Jetpack Comments iframes, tweak the iframe's parameters to send and receive `$token_key` hints about which blog token to use.

### Testing instructions:

#### Automated Tests

`phpunit --filter=WP_Test_Jetpack_Data`

#### Manual Tests

1. Get the blog token value by doing:
   `wp jetpack options get blog_token`
2. In your wp-config.php file:
   `define( 'JETPACK_BLOG_TOKEN', '{that blog token value you just found}' );`
3. `wp jetpack options delete blog_token`
   You now have a site with no stored blog token. Instead, the Jetpack will use the blog token you defined in the `JETPACK_BLOG_TOKEN` constant.
4. Go to wp-admin/ → Tools → Site Health → Info
5. Click "Jetpack"
6. Scroll down and see that "Blog Public Key" and "User Public Key" are set and that the Blog Public Key matches the token you found in step 1.


##### Testing XML-RPC & JSON API
1. Blog Token Test:
   `curl -i 'https://public-api.wordpress.com/rest/v1/sites/{YOUR_JETPACK_SITE}'`
2. User Token Test:
   1. Go to https://developer.wordpress.com/docs/api/console/
   2. Make sure you're logged in to WordPress.com as a user that corresponds to a connected user on the Jetpack site.
   3. Select "WP.COM API", "v1.1", "GET"
   4. `/sites/{YOUR_JETPACK_SITE}`
   5. Hit enter
3. For both, you should see some a JSON response of site data.

##### Testing OAuth2 Connections
1. Create an OAuth2 app and make sure you can obtain an OAuth2 token for a Jetpack site.

This is super annoying to test unless you already have a working app. Automatticians can find one at p9dueE-MU-p2.

##### Testing Jetpack Comments
1. Make sure you have Jetpack Comments turned on (wp-admin/ → Jetpack → Settings → Discussion → "Let readers use WordPress.com…"
2. Go to a post's permalink.
3. See that the Comments iframe displays correctly.
4. Submit a comment.
5. See that the comment ends up in the Jetpack site's comments at wp-admin/ → Comments.

### Proposed changelog entry for your changes:
None required.

mdawaffe committed:
Connection: DRY up Token Key Verification

When getting a Blog Token from `Jetpack_Data::get_access_token()`,
optionally double-check the retreived token matches the expected token
key.

In addition to an extra safety check, this also allows us to support
multiple blog tokens per site for some esoteric use cases.

mdawaffe committed:
Always get Blog Token via `Jetpack_Data`

Stop using `Jetpack_Options::get_option( 'blog_token' )`.

Always use `Jetpack_Data::get_access_token()`.

mdawaffe committed:
Connection: Allow Blog token to be defined by a constant

Stored token continues to function. Pick which token via token key.

mdawaffe committed:
Jetpack Comments: Send (signed) Token Key hint to WordPress.com

mdawaffe committed:
More control over what token to get.

mdawaffe committed:
Support multiple tokens in `JETPACK__BLOG_TOKENS` to better accomodate key rotation.

mdawaffe committed:
Jetpack Comments: Use (signed) Token Key Hint provided by WordPress.com

mdawaffe committed:
Connection: Rename blog token constant for better interoperability. Some external systems don't like `__`.

mdawaffe commented:
I'm adding this to 7.4 just so it gets some attention :) (Sorry!)

We'll need to ship this to Atomic soon, but it doesn't need to be part of a regular release immediately if we don't make it in time.

mdawaffe committed:
Debugger: Typo

mdawaffe committed:
Tokens: Use `Jetpack_Constants` to make the code testable.

mdawaffe committed:
Tests for `Jetpack_Data::get_access_token()`

mdawaffe committed:
XML-RPC Server: Fix verification details access during JSON API call.

PHP <= 5.3 compat.

> Fatal Error: Cannot access private property Jetpack::$xmlrpc_verification

:)

mdawaffe committed:
Tests: Oops! `->setup()` should call `parent::setup()`

jeherve commented on modules/comments/comments.php:
```diff
@@ -465,13 +476,17 @@ public function pre_comment_on_post( $comment ) {
-		$check = Jetpack_Comments::sign_remote_comment_parameters( $post_array, Jetpack_Options::get_option( 'blog_token' ) );
+		$blog_token = Jetpack_Data::get_access_token( false, $post_array['token_key'] );
+		if ( ! $blog_token ) {
+			wp_die( __( 'Unknown security token.', 'jetpack' ), 400 );
```
I seem to get stuck here in my tests. `$blog_token` returns `false` because the `hash_equals` check in `class.jetpack-data.php` fails.

jeherve reviewed (commented):
I ran into a few issues in my tests (on a site with 2 linked admin users); see below.

mdawaffe commented on class.jetpack-data.php:
```diff
@@ -26,12 +31,43 @@ public static function get_access_token( $user_id = false ) {
+				return false;
+			}
+
+			$tokens[] = $token;
```
We're interested in the first token that matches the provided `$token_key`, which is usually the first one (since most sites will only have one token), but sometimes not.

mdawaffe commented on class.jetpack-data.php:
```diff
@@ -26,12 +31,43 @@ public static function get_access_token( $user_id = false ) {
+
+			$valid_token = false;
+			foreach ( $tokens as $token ) {
+				if ( hash_equals( substr( $token, 0, strlen( $token_check ) ), $token_check ) ) {
```
What's the value of `$token_check` when this line fails?

jeherve replied:
I can't reproduce anymore, on the same site as earlier :\

mdawaffe replied:
Hm. Was it breaking for all tests (automatic, all manual ones), or just certain tests?

jeherve replied:
It was breaking when I was doing requests to `/sites/$site` via the developer console.

mdawaffe replied:
I think this was fixed in:
* b8ed3ad9020772f2664e6f51ef07804d3624eb98: never add `false` to the list of `$possible_tokens`
* 820dca9513fee0beacb04534ad232931d547c00d: allow `JETPACK_BLOG_TOKEN` to define a "Normal" (non-"Special") token. (Normal ones are… normal, Special ones are Atomic tokens, for example: the reason we're doing this in the first place :) The original code did not anticipate normal tokens in the constant, because it conflated Normal/Special v. Defined/Stored.

mdawaffe committed:
Tokens: Rename variables and simplify logic.

Rename variables to clarify what each represents.

Simplify the `$valid_token` ("this is the token to use") logic.

mdawaffe committed:
Tokens: Simplify stored token logic.

mdawaffe committed:
Tokens: Document types of tokens and magicness of magic token key.

mdawaffe committed:
Tokens: Clarify Normal v. Special.

The new Tokens code does two things:
1. Allows tokens to be stored either in a new constant or in the usual Jetpack option.
2. Differentiates between new "Special" tokens and the usual "Normal" tokens.

Earlier code conflated these two things. This commit clarifies the situation and adds
more tests: mostly around what should happen if there is a Normal token stored in the
new constant, and if there is/isn't something stored in the usual Jetpack option.

mdawaffe committed:
Tokens: Rename the Magic Normal Token Key.

`;stored;` conflated the token type (Normal v. Special) with the storage
method (Option v. Constant).

`;normal;` reduces that resulting confusion.

jeherve reviewed (approved):
This seems to work well for me now. Let's get this merged to get everyone to test it!

gravityrail commented on modules/comments/comments.php:
```diff
@@ -276,7 +276,25 @@ public function comment_form_after() {
+		list( $token_key ) = explode( '.', $blog_token->secret, 2 );
+		// Prophylactic check: anything else should never happen.
+		if ( $token_key && $token_key !== $blog_token->secret ) {
+			if ( preg_match( '/^;.\d+;\d+;$/', $token_key, $matches ) ) {
```
This regex only matches `;.(only-numbers);(only-numbers);`, but you defined this format earlier as:

`";{$special_id};.{$random_string}"`

These formats seem different.

mdawaffe replied:
The full Token Secret (the string we think of as "the token") is:

`{$token_key}.{$private}`

For Normal Tokens, both `$token_key` and `$private` are opaque random strings.

For Special Tokens, `$private` is an opaque random string, but `$token_key` looks like:
`";{$special_id};"`

where

`$special_id` looks like:

`.(only-numbers);(only-numbers)`

So `";{$special_id};.{$random_string}"` is the format of the full Special Token, and `";{$special_id};"` matches the Special Token Key format: `;.\d+;\d+;`.

Sounds like I should clean up the docs here :)

gravityrail replied:
Thank you for the explanation!
